src/Widgets/PartitionMenu.vala

```diff
@@ -0,0 +1,355 @@
+        label_size_group.add_widget (custom_label);
+
+        custom = new Gtk.Entry ();
+
+        type_label = new Gtk.Label ("Filesystem:");
+        type_label.halign = Gtk.Align.END;
+        type_label.xalign = 1;
+        label_size_group.add_widget (type_label);
+
+        type = new Gtk.ComboBoxText ();
+        type.append_text (_("Default (ext4)"));
+        type.append_text ("fat16");
+        type.append_text ("fat32");
+        type.append_text ("btrfs");
+        type.append_text ("xfs");
+        type.append_text ("ntfs");
+        type.active = 0;
+
+        var top_controls = new Gtk.Grid ();
+        top_controls.column_spacing = 12;
+        top_controls.row_spacing = 6;
+        top_controls.margin = 12;
+
+        top_controls.attach (use_partition_label, 0, 0);
+        top_controls.attach (use_partition,       1, 0);
+
+        var bottom_controls = new Gtk.Grid ();
+        bottom_controls.column_spacing = 12;
+        bottom_controls.row_spacing = 6;
+        bottom_controls.margin = 12;
+        bottom_controls.margin_top = 6;
+
+        bottom_controls.attach (format_label, 0, 1);
+        bottom_controls.attach (use_as_label, 0, 2);
+        bottom_controls.attach (custom_label, 0, 3);
+        bottom_controls.attach (type_label,   0, 4);
+
+        bottom_controls.attach (format_partition, 1, 1);
+        bottom_controls.attach (use_as,           1, 2);
+        bottom_controls.attach (custom,           1, 3);
+        bottom_controls.attach (type,             1, 4);
+
+        var bottom_grid = new Gtk.Grid ();
+        bottom_grid.column_spacing = 12;
+        bottom_grid.row_spacing = 6;
+
+        bottom_grid.attach (separator,       0, 0);
+        bottom_grid.attach (bottom_controls, 0, 1);
+
+        var grid = new Gtk.Grid ();
+        grid.column_spacing = 12;
+
+        var bottom_revealer = new Gtk.Revealer ();
+        bottom_revealer.add (bottom_grid);
+
+        grid.attach (top_controls,    0, 0);
+        grid.attach (bottom_revealer, 0, 1);
+
+        add (grid);
+        grid.show_all ();
+
+        // FIXME: Probably should stick these in a revealer, too
+        custom.visible = false;
+        custom_label.visible = false;
+
+        format_partition.notify["active"].connect (() => {
+            if (!disable_signals) {
+                check_values (set_mount);
+            }
+        });
+
+        use_as.changed.connect (() => {
+            if (disable_signals) {
+                return;
+            }
+
+            var active = use_as.active;
+            bool visible = active == 4;
+
+            custom.visible = visible;
+            custom_label.visible = visible;
+
+            if (active == 2) {
+                if (Distinst.bootloader_detect () == Distinst.PartitionTable.GPT) {
+                    type.active = 2;
+                } else {
+                    type.active = 0;
+                }
+                type_label.visible = true;
+                type.visible = true;
```

Comment:
you might want to use `bind_property` between `type` and `type_label` at creation time as they seem to share their visible property and I think that there are others here. it could just make it a little bit more lightweight